app/portainer/components/endpoint-list/endpoint-item/endpoint-item-controller.js

```diff
@@ -1,12 +1,39 @@
-angular.module('portainer.app').controller('EndpointItemController', [
-  function EndpointItemController() {
-    var ctrl = this;
+angular.module('portainer.app').controller('EndpointItemController', function EndpointItemController() {
+  var ctrl = this;
+  this.endpointTags = '';
 
-    ctrl.editEndpoint = editEndpoint;
+  ctrl.editEndpoint = editEndpoint;
 
-    function editEndpoint(event) {
-      event.stopPropagation();
-      ctrl.onEdit(ctrl.model.Id);
+  this.$onInit = function $onInit() {
+    this.endpointTags = this.joinTags();
+  };
+  this.$onChanges = function $onChanges(changesObj) {
+    this.handleTagsChange(changesObj);
+  };
+  this.handleTagsChange = function handleTagsChange({ tags, model }) {
+    if ((!tags && !model) || (!tags.currentValue && !model.currentValue)) {
+      return;
     }
+    this.endpointTags = this.joinTags();
+  };
+
+  function editEndpoint(event) {
+    event.stopPropagation();
+    ctrl.onEdit(ctrl.model.Id);
+  }
+
+  this.joinTags = function joinTags() {
+    if (!this.tags) {
+      return 'Loading tags...'
+    }
+    return this.model.TagIds.map(tagId => {
```

Comment:
I would advocate against returning a function for clarity/readility.

Prepare the variable, assign it to your function and then returns the function. Shorter code do not make it easier to maintain. Clear code will.

Reply:
not sure what you mean, something like this?
```
function findTagName(tag) {
  ...
}
return this.model.TagIds.map(findTagName);
```

Reply:
the advantage of using arrow function is that the binding of `this` doesn't change. with this solution, we would need to rebind `this`

Reply:
please have a look at my fix